godot-ffi/src/godot_ffi.rs

```diff
@@ -129,94 +134,12 @@ pub unsafe fn from_sys_init_or_init_default<T: GodotFfi>(
-/// compatible types, without knowing their definition.
-///
-/// # Safety
+/// Types that can represent null-values.
 ///
-/// The type has to have a pointer-sized counterpart in Godot, which needs to be nullable.
-/// So far, this only applies to class types (Object hierarchy).
-pub unsafe trait GodotNullablePtr: GodotFfi {}
-
-unsafe impl<T> GodotFfi for Option<T>
-where
-    T: GodotNullablePtr,
-{
-    unsafe fn from_sys(ptr: sys::GDExtensionTypePtr) -> Self {
-        crate::ptr_then(ptr, |ptr| T::from_sys(ptr))
-    }
-
-    unsafe fn from_sys_init(init_fn: impl FnOnce(sys::GDExtensionUninitializedTypePtr)) -> Self {
-        let mut raw = std::mem::MaybeUninit::uninit();
-        init_fn(raw.as_mut_ptr() as sys::GDExtensionUninitializedTypePtr);
-
-        Self::from_sys(raw.assume_init())
-    }
-
-    fn sys(&self) -> sys::GDExtensionTypePtr {
-        match self {
-            Some(value) => value.sys(),
-            None => ptr::null_mut() as sys::GDExtensionTypePtr,
-        }
-    }
-
-    #[cfg(before_api = "4.1")]
-    unsafe fn from_arg_ptr(ptr: sys::GDExtensionTypePtr, call_type: PtrcallType) -> Self {
-        match call_type {
-            PtrcallType::Standard => option_from_arg_single_ptr(ptr, call_type),
-            PtrcallType::Virtual => option_from_arg_double_ptr(ptr, call_type),
-        }
-    }
-
-    #[cfg(since_api = "4.1")]
-    unsafe fn from_arg_ptr(ptr: sys::GDExtensionTypePtr, call_type: PtrcallType) -> Self {
-        option_from_arg_double_ptr(ptr, call_type)
-    }
-
-    unsafe fn move_return_ptr(self, ptr: sys::GDExtensionTypePtr, call_type: PtrcallType) {
-        if let Some(value) = self {
-            value.move_return_ptr(ptr, call_type)
-        }
-    }
-}
-
-/// Return an `Option<T>` when `T` is a nullable pointer type and `ptr` is represented as `T**`.
-///
-/// # Safety
-/// `ptr` must either be null, a pointer to null, or a pointer to a pointer to a `T`.
-unsafe fn option_from_arg_double_ptr<T>(
-    ptr: sys::GDExtensionTypePtr,
-    call_type: PtrcallType,
-) -> Option<T>
-where
-    T: GodotNullablePtr,
-{
-    if ptr.is_null() || (*(ptr as *mut sys::GDExtensionTypePtr)).is_null() {
-        None
-    } else {
-        Some(T::from_arg_ptr(ptr, call_type))
-    }
-}
-
-// 4.1 represents every object as `T**`.
-#[cfg(before_api = "4.1")]
-/// Return an `Option<T>` when `T` is a nullable pointer type and `ptr` is represented as `T*`.
-///
-/// # Safety
-/// `ptr` must either be null, or a pointer to a `T`.
-unsafe fn option_from_arg_single_ptr<T>(
-    ptr: sys::GDExtensionTypePtr,
-    call_type: PtrcallType,
-) -> Option<T>
-where
-    T: GodotNullablePtr,
-{
-    crate::ptr_then(ptr, |ptr| T::from_arg_ptr(ptr, call_type))
+/// Used to blanket implement various conversions over `Option<T>`.
+pub trait GodotNullableFfi: Sized + GodotFfi {
+    fn flatten_option(opt: Option<Self>) -> Self;
+    fn is_null(&self) -> bool;
 }
```

Comment:
I would also mention for which (kinds of) types this trait is implemented.

A user probably won't need to face this directly?

Reply:
Yeah they shouldn't. We could consider implementing this for `Variant`, since they have a meaningful null-state. But for now it's only implemented for `RawGd`, since that's the least amount of changes from the current behavior.